BoundedKnapsack/ReferenceImplementation.qs

```diff
@@ -0,0 +1,409 @@
+        // In the future, this will be replaced by the quantum counting algorithm.
+        let N = IntAsDouble(1 <<< Q);
+        let m = IntAsDouble(NumberOfSolutions(n, W, P, itemWeights, itemProfits, itemInstanceBounds));
+        if (m == 0.0){
+            return (xs_found, P_found);
+        }
+        // Using the formula for the number of iterations, and rounding to the nearest integer
+        mutable iter = Floor(PI() / 4.0 * Sqrt(N/m)+0.5);
+        mutable attempts = 0;
+
+        using (register = Qubit[Q]){
+            
+            repeat {
+                // Note: The register is not converted into the jagged array before being used in the oracle, because
+                //         the ApplyToEach operations in the GroverIterations can't directly be called on jagged arrays.
+                GroversAlgorithm_Loop(register, KnapsackValidationOracle_Reference(n, W, P, maxTotal, itemWeights, itemProfits, itemInstanceBounds, _, _), iter);
+
+                // Measure the combination that Grover's Algorithm finds.
+                let xs = RegisterAsJaggedArray_Reference(n, itemInstanceBounds, register);
+                for (i in 0..n-1){
+                    let result = MultiM(xs[i]);
+                    set xs_found w/= i <- ResultArrayAsInt(result);
+                }
+
+                // Check that the combination is a valid combination.
+                using (output = Qubit()){
+                    KnapsackValidationOracle_Reference(n, W, P, maxTotal, itemWeights, itemProfits, itemInstanceBounds, register, output);
+                    if (MResetZ(output) == One){
+                        set correct = true;
+                    }
+                }
+
+                // When the valid combination is found, calculate its profit
+                if (correct){
+                    using (profit = Qubit[maxTotal]){
+                        CalculateTotalWeightOrProfit_Reference(n, itemProfits, xs, profit);
+                        set P_found = ResultArrayAsInt(MultiM(profit));
+                        ResetAll(profit);
+                    }
+                }
+                ResetAll(register);
+                set attempts += 1;
+            } until(correct or attempts > 10);
+
+        }
+
+        return (xs_found, P_found);
+        
+    }
+
+    function RegisterSize(n : Int, itemInstanceBounds : Int[]) : Int {
+        // Calculate the total number of qubits for the register, given the bounds array. The item with index i can have 0 to bᵢ instances,
+        // which requires log₂(bᵢ+1) qubits (rounded up). The auxiliary function BitSizeI is used to faciliate
+        // this calculation. The total number of qubits, Q, is the sum of each individual number of qubits.
+        mutable Q = 0;
+        for (bound in itemInstanceBounds){
+            set Q += BitSizeI(bound);
+        }
+        return Q;
+    }
+
+    // Grover loop implementation taken from SolveSATWithGrover kata.

```

Comment:
@tcNickolas Would it make sense to extract these functions into a separate common project?

Reply:
I agree that it would be nice to reduce the code duplication (this is not the last kata in which I expect to see Grover's search!)

Do you think [GroversAlgorithm](https://github.com/microsoft/QuantumKatas/tree/master/GroversAlgorithm) could work as such a project, or do we need a new one? It is the "first time" we define this loop, and there it serves educational purpose, so I wouldn't remove the definition of the loop from there.

Reply:
@tcNickolas @msoeken I think moving these functions to [GroversAlgorithm](https://github.com/microsoft/QuantumKatas/tree/master/GroversAlgorithm) would be a very good idea, although I'm not quite sure how to implement it. Would this project and future projects that use Grover be able to call those functions despite being in a different project, or would GroversAlgorithm have to be imported?
In addition, by "these functions", are you referring to only OracleConverterImpl and GroversAlgorithm_Loop, or also the GroversAlgorithm itself? It would make sense to me if the former two were extracted, but not GroversAlgorithm, since each kata implements Grover's Algorithm differently (i.e. whether it returns a bool, or the found combination, etc.)

Reply:
We should have a separate PR for this discussion. Then we can investigate how much the operations differ and whether it is worth to generalise some. But I think it's good to have them in the GroversAlgorithm project.